Show Problem Types In Contest When Enabled

tandr3w opened the pull request:
Fixes #1872 (or attempts to)

int-y1 commented on judge/views/problem.py:
```diff
@@ -490,7 +492,9 @@ def get_queryset(self):
     def get_context_data(self, **kwargs):
         context = super(ProblemList, self).get_context_data(**kwargs)
         context['hide_solved'] = 0 if self.in_contest else int(self.hide_solved)
-        context['show_types'] = 0 if self.in_contest else int(self.show_types)
```
why did you avoid
```py
context['show_types'] = int(not self.contest.hide_problem_tags) if self.in_contest else int(self.show_types)
```

tandr3w replied:
> why did you avoid
> 
> ```python
> context['show_types'] = int(not self.contest.hide_problem_tags) if self.in_contest else int(self.show_types)
> ```

I wasn't sure whether or not the self.show_types attribute was used elsewhere and had to be changed as well

int-y1 replied:
`get_context_data` doesn't look like the right place to modify `self.show_types`. what happens if you put your changes in `setup_problem_list`?

int-y1 commented on judge/views/problem.py:
```diff
@@ -425,8 +426,9 @@ def get_contest_queryset(self):
-        } for p in queryset.values('problem_id', 'problem__code', 'problem__name', 'i18n_name',
-                                   'problem__group__full_name', 'points', 'partial', 'user_count')]
+            'types_list': q.problem.types_list,
+        } for p, q in zip(queryset.values('problem_id', 'problem__code', 'problem__name', 'i18n_name',
+                                          'problem__group__full_name', 'points', 'partial', 'user_count'), queryset)]
```
`zip(queryset.values(...), queryset)` is pretty weird, is there a better way to do this?

int-y1 replied:
`zip(queryset.values(...), queryset)` defeats the point of having a `queryset.values(...)`

tandr3w replied:
> `zip(queryset.values(...), queryset)` is pretty weird, is there a better way to do this?

Couldn't think of a solution using queryset.values(...) since I believe it can't access the types_list property. Maybe getting rid of the queryset.values(...) would be better?

quantum5 replied:
The `zip` is fundamentally broken.

int-y1 commented:
can you post screenshots of the problem list?

tandr3w commented:
Before:
![image](https://github.com/DMOJ/online-judge/assets/84942946/5985b029-ed7d-44a2-8051-e1a7ec2576ce)

After: 
![image](https://github.com/DMOJ/online-judge/assets/84942946/0474aa37-0598-4fd0-8f41-7f10599b5b0c)

quantum5 reviewed (changes requested):
PR is completely borked, not to mention inefficient.

tandr3w commented:
- Moved show_types change into the setup_problem_list function
- Replaced zip with an annotation
- Fixed inconsistent padding due to the in-contest types table header not having the "types" class

tandr3w commented:
New Screenshots:

Before changes:
![image](https://github.com/DMOJ/online-judge/assets/84942946/bade812e-d1ee-4292-94a1-791d6860c1d5)

With types disabled, after changes
![image](https://github.com/DMOJ/online-judge/assets/84942946/32594711-df6f-434f-abcf-7b0a5dcc6259)

With types enabled, after changes
![image](https://github.com/DMOJ/online-judge/assets/84942946/db38b431-6806-48ac-9baa-fdf5dea41228)

tandr3w commented:
Now works for translated types, ready for another review
![image](https://github.com/DMOJ/online-judge/assets/84942946/3150c313-7fd9-4a39-8b73-742d856adb8d)

tandr3w commented:
Ready for re-review
![image](https://github.com/DMOJ/online-judge/assets/84942946/8fbcf358-c73e-445b-9c58-68335ba27c7f)
